pydantic/env_settings.py

```diff
@@ -170,25 +172,70 @@ def __call__(self, settings: BaseSettings) -> Dict[str, Any]:
                 continue
 
             if field.is_complex():
-                try:
-                    env_val = settings.__config__.json_loads(env_val)
-                except ValueError as e:
-                    raise SettingsError(f'error parsing JSON for "{env_name}"') from e
+                if isinstance(env_val, (str, bytes, bytearray)):
+                    try:
+                        env_val = settings.__config__.json_loads(env_val)
+                    except ValueError as e:
+                        raise SettingsError(f'error parsing JSON for "{env_name}"') from e
             elif (
                 is_union_origin(get_origin(field.type_))
                 and field.sub_fields
                 and any(f.is_complex() for f in field.sub_fields)
             ):
-                try:
-                    env_val = settings.__config__.json_loads(env_val)
-                except ValueError:
-                    pass
+                if isinstance(env_val, (str, bytes, bytearray)):
+                    try:
+                        env_val = settings.__config__.json_loads(env_val)
+                    except ValueError:
+                        pass
             d[field.alias] = env_val
         return d
 
     def __repr__(self) -> str:
         return f'EnvSettingsSource(env_file={self.env_file!r}, env_file_encoding={self.env_file_encoding!r})'
 
+    def get_env_vars(self, settings: BaseSettings) -> Mapping[str, Optional[str]]:
+        if settings.__config__.case_sensitive:
+            env_vars: Mapping[str, Optional[str]] = os.environ
+        else:
+            env_vars = {k.lower(): v for k, v in os.environ.items()}
+
+        if self.env_file is not None:
+            env_path = Path(self.env_file).expanduser()
+            if env_path.is_file():
+                env_vars = {
+                    **read_env_file(
+                        env_path, encoding=self.env_file_encoding, case_sensitive=settings.__config__.case_sensitive
+                    ),
+                    **env_vars,
+                }
+        return env_vars
+
+    def explode_env_vars(self, settings: BaseSettings, env_vars: Mapping[str, Optional[str]]) -> Dict[str, Any]:
+        """
+        Go trough environment variables.
+        Within the each key-value pair:
+            - Split key with the env_nested_delimiter
+            - Before going through each key define env_var as a root (same as the final result)
+            - Go trough each key, add a new dict item to the env_var and redefine env_var within this item
+            - At the last iteration:
+                - Silently try to parse value with the json, leave it as it is if it fails
+                - Assign the value to the deepest key
+        """
+        result: Dict[str, Any] = {}
+        for env_name, env_val in env_vars.items():
+            keys = env_name.split(self.env_nested_delimiter)
+            env_var = result
```

Comment:
I don't think this approach works.

If you change the order of env variables in your test below, the result changes. E.g. put `env.set('top__sub__v5', '5')` first.

Reply:
Thank you for pointing on this. Yes there was a bug. Redesigned this method and test a bit according to this bug and logic from my answer on your comment above.